### ISSUE TITLE
Freezing layers for fine tuning

### ISSUE THREAD

**Issue by aman313** [allenai/allennlp#1298]
Hi folks,

Thanks for this excellent library. I am trying to use the Bidaf model in a few shot setting where I want to train the model on larger dataset and then freeze the initial layers and retrain the output layer on a smaller dataset. I have been unable to find an easy way for doing this with the library yet. Am I missing something? 

Cheers,
Aman

**Comment by aman313** [comment]
I managed to get this done. Answering my own question for any suggested improvements or for the benefit of someone else looking to do this.
1. Create a new registrable model ("bidaf_tunable") which is a copy of the bidaf code.
2. Changed the from_params function to accept a new param "frozen" which is list of layer names which should be frozen. Set requires_grad for the layer params as false.
3. used the fine-tune command to load model from old weights and train on new data.

**Comment by matt-gardner** [comment]
@aman313, nice job fixing the issue, that's a good way to do it with our current code (though pretty unfortunate that you need to copy the whole model...).  Let us think about a better way to make this functionality available more generally.

Maybe it makes sense, for instance, to put an option on the `fine-tune` command itself, that lets you freeze layers?  I can't think of a good way to have this just work for all models using `train`, without requiring all models to take an additional parameter.

**Comment by aman313** [comment]
@matt-gardner Yeah. The fine tune command is the first place I looked too :). 
If I understand you (and the code!) correctly, the option in fine tune command could load all weights and then freeze certain params before calling train? For the train command I assume you mean restarting training with --recover option? We could change the code to freeze params here too but I think recovering the training state, specifically the optimizer, might cause some inconsistency. We could add another option to disallow loading some objects from the serialized training state? Does that make sense?

**Comment by matt-gardner** [comment]
For using `train`, I was thinking of models that are at least partially initialized with some other (non-identical) model.  Those models can probably figure out their own means of freezing state, so after some more thought I don't actually think this notion is applicable to `train` at all.

And yes, for `fine-tune`, there would be a flag on the main command that listed a set modules or parameter names.  After the model is loaded, you would loop over `model.named_parameters()` and set `requires_grad=False` to all of those in the given list.  Something like this _should_ work, though I might have some details wrong.

**Comment by HarshTrivedi** [comment]
It can be done with prevent initialization regex as well (#1405). 

Lets consider two scenarios in which one can transfer weights:

1. model is exactly same but you want to freeze some parameters on continuing training. This is the scenario discussed above. Eg. I trained model with dataA, now want to freeze "linear_x_*" and continue other parameters training on dataB.
2. model itself is new and more complex. I trained modelA on dataA, but now I want to use modelA's trained encoder_x and encoder_y in my new modelB. Moreover, I want to freeze encoder_x and keep encoder_y learnable. This can also be extended for borrowing modules from different trained models.

**Scenario 1** is simpler and should be easily handlable with fine-tune command. Only minor change in fine-tune is required to turn of gradients of relevant parameters. It would be better if regex based support was available in experiment.json to control this. I can send pr for this after discussion here.

**Scenario 2**: With (#1405), there is a support for scenario 2 using `train` command. Below is a bare-minimum sketch of `modelB.py`.

```
def from_params(cls, vocab: Vocabulary, params: Params):
    modelA = load_archive( params.pop("modelA_path")).model
    encoder_x = modelA.encoder_x
    encoder_y = modelA.encoder_y
    initializer_params = params.pop('initializer', [])
    initializer_params += [ [".*encoder_x.*".format(name), "prevent"]  ]  # because it needs to be transferred
    initializer_params += [ [".*encoder_y.*".format(name), "prevent"]  ]  # because it needs to be transferred
    initializer = InitializerApplicator.from_params(initializer_params)

    nograd_module_names = ["encoder_x"]

    return cls(vocab=vocab,
                     encoder_x=encoder_x,
                     encoder_y=encoder_y,
                     nograd_module_names=nograd_module_names)

def __init__(....):

    # do everyting ..

    initializer(self)
    for module_name in nograd_module_names:
        [ p.requires_grad_(False) for p in self._modules[module_name].parameters() ]

```

I have written some utilities to handle more complicated versions of Scenario 2 directly with `experiment.json` so one can do the following:
```
"model":{
        "modelA": {
            "dir": "modelA_ser_dir",
            "borrow_modules": { 
                    "encoder_x" : "transfer" # Only Transfer
                    "encoder_y" : "tune"     # Transfer and Tune
            }
        },
}
```
and have the above effect with 2-3 additional lines using utilities in `from_params`. These utilities were written for my specific purpose because I am borrowing parts and pieces from different models at the same time and need full configurability. But if the usage seems general enough, we can discuss this in separate pr.
